File: examples/11_numerics/compare/semantics.c
/* The same three questions put to C: signed overflow is undefined, division truncates. */
#include <stdio.h>

static long long may_overflow(long long n) {
    long long result = 1;
    for (long long i = 1; i <= n; i++) result *= i;
    return result;
}

static long long floor_semantics(long long a, long long b) { return a / b; }
static long long modulo_semantics(long long a, long long b) { return a % b; }

int main(void) {
    printf("%lld\n", may_overflow(20));
    printf("%lld\n", may_overflow(30));
    printf("%lld %lld\n", floor_semantics(-7, 2), floor_semantics(7, 2));
    printf("%lld %lld\n", modulo_semantics(-7, 2), modulo_semantics(7, -2));
    return 0;
}

File: examples/11_numerics/compare/semantics.rs
//! The same three questions put to Rust: `i64` panics on overflow in debug, wraps in release.

fn may_overflow(n: i64) -> i64 {
    let mut result: i64 = 1;
    for i in 1..=n {
        result *= i;
    }
    result
}

fn floor_semantics(a: i64, b: i64) -> i64 {
    a / b
}

fn modulo_semantics(a: i64, b: i64) -> i64 {
    a % b
}

fn main() {
    println!("{}", may_overflow(20));
    println!("{}", may_overflow(30));
    println!("{} {}", floor_semantics(-7, 2), floor_semantics(7, 2));
    println!("{} {}", modulo_semantics(-7, 2), modulo_semantics(7, -2));
}
